File: frontend/src/lib/analytics.ts
/**
 * Analytics Service
 * Handles data fetching and processing for analytics dashboard
 */

export interface DateRange {
  start_date?: string;
  end_date?: string;
}

export interface RevenueAnalytics {
  total_revenue: number;
  average_order_value: number;
  revenue_growth_percentage: number;
  revenue_by_payment_method: Array<{
    payment_method: string;
    revenue: number;
    order_count: number;
  }>;
  daily_revenue_trend: Array<{
    date: string;
    revenue: number;
    orders: number;
  }>;
  period: {
    start_date: string;
    end_date: string;
  };
}

export interface CustomerAnalytics {
  total_customers: number;
  new_customers: number;
  active_customers: number;
  customer_lifetime_value: number;
  top_customers: Array<{
    id: number;
    email: string;
    full_name: string;
    total_spent: number;
    order_count: number;
  }>;
  customer_acquisition_trend: Array<{
    year: number;
    month: number;
    new_customers: number;
  }>;
}

export interface ProductAnalytics {
  top_selling_products: Array<{
    id: number;
    name: string;
    price: number;
    total_sold: number;
    total_revenue: number;
    order_count: number;
  }>;
  category_performance: Array<{
    category: string;
    total_sold: number;
    total_revenue: number;
    product_count: number;
  }>;
  conversion_rates: any[];
}

export interface PaymentAnalytics {
  payment_method_distribution: Array<{
    payment_method: string;
    transaction_count: number;
    total_amount: number;
    average_amount: number;
  }>;
  success_rates: Array<{
    payment_method: string;
    successful_transactions: number;
    failed_transactions: number;
    total_transactions: number;
    success_rate: number;
  }>;
  daily_transaction_volume: Array<{
    date: string;
    transaction_count: number;
    total_amount: number;
  }>;
}

export interface ExecutiveSummary {
  summary: {
    total_revenue: number;
    total_customers: number;
    new_customers: number;
    average_order_value: number;
    revenue_growth: number;
  };
  top_metrics: {
    best_selling_product: any;
    preferred_payment_method: any;
    top_customer: any;
  };
  trends: {
    revenue_trend: any[];
    customer_acquisition: any[];
  };
  period: {
    start_date: string;
    end_date: string;
  };
}

export interface RealTimeMetrics {
  today_revenue: number;
  yesterday_revenue: number;
  revenue_change: number;
  current_hour_orders: number;
  active_sessions: number;
  conversion_rate: number;
  last_updated: string;
}

export class AnalyticsService {
  private baseUrl: string;
  private getAuthHeaders: () => Record<string, string>;

  constructor(baseUrl?: string) {
    this.baseUrl = baseUrl || process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
    this.getAuthHeaders = () => {
      const token = typeof window !== 'undefined' ? localStorage.getItem('authToken') : '';
      return {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      };
    };
  }

  private async fetchWithAuth(endpoint: string, options?: any) {
    const response = await fetch(`${this.baseUrl}${endpoint}`, {
      ...options,
      headers: {
        ...this.getAuthHeaders(),
        ...options?.headers,
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    
    if (!data.success) {
      throw new Error(data.message || 'API request failed');
    }

    return data.data;
  }

  private buildQueryParams(params: DateRange): string {
    const searchParams = new URLSearchParams();
    
    if (params.start_date) {
      searchParams.append('start_date', params.start_date);
    }
    
    if (params.end_date) {
      searchParams.append('end_date', params.end_date);
    }

    const queryString = searchParams.toString();
    return queryString ? `?${queryString}` : '';
  }

  // Revenue Analytics
  async getRevenueAnalytics(dateRange?: DateRange): Promise<RevenueAnalytics> {
    const queryParams = this.buildQueryParams(dateRange || {});
    return this.fetchWithAuth(`/api/v1/analytics/revenue${queryParams}`);
  }

  // Customer Analytics
  async getCustomerAnalytics(dateRange?: DateRange): Promise<CustomerAnalytics> {
    const queryParams = this.buildQueryParams(dateRange || {});
    return this.fetchWithAuth(`/api/v1/analytics/customers${queryParams}`);
  }

  // Product Analytics
  async getProductAnalytics(dateRange?: DateRange): Promise<ProductAnalytics> {
    const queryParams = this.buildQueryParams(dateRange || {});
    return this.fetchWithAuth(`/api/v1/analytics/products${queryParams}`);
  }

  // Payment Analytics
  async getPaymentAnalytics(dateRange?: DateRange): Promise<PaymentAnalytics> {
    const queryParams = this.buildQueryParams(dateRange || {});
    return this.fetchWithAuth(`/api/v1/analytics/payments${queryParams}`);
  }

  // Executive Dashboard
  async getExecutiveDashboard(dateRange?: DateRange): Promise<ExecutiveSummary> {
    const queryParams = this.buildQueryParams(dateRange || {});
    return this.fetchWithAuth(`/api/v1/analytics/dashboard${queryParams}`);
  }

  // Real-time Metrics
  async getRealTimeMetrics(): Promise<RealTimeMetrics> {
    return this.fetchWithAuth('/api/v1/analytics/real-time');
  }

  // Export Report
  async exportReport(
    reportType: 'revenue' | 'customers' | 'products' | 'payments',
    format: 'json' | 'csv' = 'json',
    dateRange?: DateRange
  ): Promise<any> {
    const queryParams = new URLSearchParams({
      report_type: reportType,
      format,
      ...(dateRange?.start_date && { start_date: dateRange.start_date }),
      ...(dateRange?.end_date && { end_date: dateRange.end_date }),
    });

    return this.fetchWithAuth(`/api/v1/analytics/export?${queryParams}`);
  }

  // Get All Analytics Data
  async getAllAnalytics(dateRange?: DateRange) {
    const [revenue, customers, products, payments, dashboard] = await Promise.all([
      this.getRevenueAnalytics(dateRange),
      this.getCustomerAnalytics(dateRange),
      this.getProductAnalytics(dateRange),
      this.getPaymentAnalytics(dateRange),
      this.getExecutiveDashboard(dateRange),
    ]);

    return {
      revenue,
      customers,
      products,
      payments,
      dashboard,
    };
  }

  // Helper method to format currency
  static formatCurrency(amount: number, currency: string = 'USD'): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency,
    }).format(amount);
  }

  // Helper method to format percentage
  static formatPercentage(value: number, decimals: number = 1): string {
    return `${value.toFixed(decimals)}%`;
  }

  // Helper method to calculate growth rate
  static calculateGrowthRate(current: number, previous: number): number {
    if (previous === 0) return current > 0 ? 100 : 0;
    return ((current - previous) / previous) * 100;
  }

  // Helper method to get period comparison
  static getPeriodComparison(current: any, previous: any) {
    return {
      revenue_growth: this.calculateGrowthRate(current.total_revenue, previous.total_revenue),
      customer_growth: this.calculateGrowthRate(current.total_customers, previous.total_customers),
      order_growth: this.calculateGrowthRate(current.total_orders, previous.total_orders),
    };
  }

  // Real-time data subscription (placeholder for WebSocket implementation)
  subscribeToRealTimeUpdates(callback: (_data: RealTimeMetrics) => void): () => void {
    // This would implement WebSocket connection for real-time updates
    const interval = setInterval(async () => {
      try {
        const data = await this.getRealTimeMetrics();
        callback(data);
      } catch (error) {
        // Failed to fetch real-time metrics
      }
    }, 30000); // Update every 30 seconds

    // Return unsubscribe function
    return () => clearInterval(interval);
  }
}

// Export singleton instance
export const analyticsService = new AnalyticsService();

// Export utility functions
export const AnalyticsUtils = {
  formatCurrency: AnalyticsService.formatCurrency,
  formatPercentage: AnalyticsService.formatPercentage,
  calculateGrowthRate: AnalyticsService.calculateGrowthRate,
  getPeriodComparison: AnalyticsService.getPeriodComparison,
};

File: backend/app/services/analytics.py
"""
Advanced Analytics Service
Provides comprehensive reporting and insights for BrainSAIT B2B platform
"""

import json
from datetime import date, datetime, timedelta
from decimal import Decimal
from typing import Any, Dict, List, Optional, Tuple

from sqlalchemy import and_, asc, case, desc, extract, func, or_
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.orders import Order, OrderItem, OrderStatus, PaymentStatus
from app.models.products import Product
from app.models.users import User


class AnalyticsService:
    """Advanced analytics and reporting service"""

    def __init__(self, db: Session):
        self.db = db

    async def get_revenue_analytics(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        tenant_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Get comprehensive revenue analytics"""

        # Default to last 30 days if no dates provided
        if not end_date:
            end_date = date.today()
        if not start_date:
            start_date = end_date - timedelta(days=30)

        # Base query filters
        filters = [
            Order.created_at >= start_date,
            Order.created_at <= end_date + timedelta(days=1),
            Order.status == OrderStatus.COMPLETED,
        ]

        if tenant_id:
            filters.append(Order.tenant_id == tenant_id)

        # Total revenue
        total_revenue = self.db.query(func.sum(Order.total_amount)).filter(
            *filters
        ).scalar() or Decimal("0.00")

        # Revenue by payment method
        revenue_by_payment = (
            self.db.query(
                Order.payment_method,
                func.sum(Order.total_amount).label("revenue"),
                func.count(Order.id).label("order_count"),
            )
            .filter(*filters)
            .group_by(Order.payment_method)
            .all()
        )

        # Daily revenue trend
        daily_revenue = (
            self.db.query(
                func.date(Order.created_at).label("date"),
                func.sum(Order.total_amount).label("revenue"),
                func.count(Order.id).label("orders"),
            )
            .filter(*filters)
            .group_by(func.date(Order.created_at))
            .order_by(func.date(Order.created_at))
            .all()
        )

        # Average order value
        avg_order_value = self.db.query(func.avg(Order.total_amount)).filter(
            *filters
        ).scalar() or Decimal("0.00")

        # Revenue growth (compared to previous period)
        prev_start = start_date - (end_date - start_date)
        prev_end = start_date

        prev_filters = [
            Order.created_at >= prev_start,
            Order.created_at <= prev_end,
            Order.status == OrderStatus.COMPLETED,
        ]

        if tenant_id:
            prev_filters.append(Order.tenant_id == tenant_id)

        prev_revenue = self.db.query(func.sum(Order.total_amount)).filter(
            *prev_filters
        ).scalar() or Decimal("0.00")

        revenue_growth = 0.0
        if prev_revenue > 0:
            revenue_growth = float((total_revenue - prev_revenue) / prev_revenue * 100)

        return {
            "total_revenue": float(total_revenue),
            "average_order_value": float(avg_order_value),
            "revenue_growth_percentage": revenue_growth,
            "revenue_by_payment_method": [
                {
                    "payment_method": payment.value if payment else "unknown",
                    "revenue": float(revenue),
                    "order_count": order_count,
                }
                for payment, revenue, order_count in revenue_by_payment
            ],
            "daily_revenue_trend": [
                {
                    "date": date_val.isoformat(),
                    "revenue": float(revenue),
                    "orders": orders,
                }
                for date_val, revenue, orders in daily_revenue
            ],
            "period": {
                "start_date": start_date.isoformat(),
                "end_date": end_date.isoformat(),
            },
        }

    async def get_customer_analytics(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        tenant_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Get customer behavior analytics"""

        if not end_date:
            end_date = date.today()
        if not start_date:
            start_date = end_date - timedelta(days=30)

        base_filters = []
        if tenant_id:
            base_filters.append(User.tenant_id == tenant_id)

        # Total customers
        total_customers = (
            self.db.query(func.count(User.id)).filter(*base_filters).scalar()
        )

        # New customers in period
        new_customers = (
            self.db.query(func.count(User.id))
            .filter(
                User.created_at >= start_date,
                User.created_at <= end_date + timedelta(days=1),
                *base_filters,
            )
            .scalar()
        )

        # Active customers (with orders in period)
        active_customers = (
            self.db.query(func.count(func.distinct(Order.user_id)))
            .join(User)
            .filter(
                Order.created_at >= start_date,
                Order.created_at <= end_date + timedelta(days=1),
                *base_filters,
            )
            .scalar()
        )

        # Customer lifetime value
        customer_ltv = self.db.query(
            func.avg(
                self.db.query(func.sum(Order.total_amount))
                .filter(Order.user_id == User.id)
                .scalar_subquery()
            )
        ).filter(*base_filters).scalar() or Decimal("0.00")

        # Top customers by revenue
        top_customers = (
            self.db.query(
                User.id,
                User.email,
                User.full_name,
                func.sum(Order.total_amount).label("total_spent"),
                func.count(Order.id).label("order_count"),
            )
            .join(Order)
            .filter(
                Order.created_at >= start_date,
                Order.created_at <= end_date + timedelta(days=1),
                Order.status == OrderStatus.COMPLETED,
                *base_filters,
            )
            .group_by(User.id, User.email, User.full_name)
            .order_by(desc("total_spent"))
            .limit(10)
            .all()
        )

        # Customer acquisition by month
        customer_acquisition = (
            self.db.query(
                extract("year", User.created_at).label("year"),
                extract("month", User.created_at).label("month"),
                func.count(User.id).label("new_customers"),
            )
            .filter(*base_filters)
            .group_by(
                extract("year", User.created_at), extract("month", User.created_at)
            )
            .order_by("year", "month")
            .all()
        )

        return {
            "total_customers": total_customers,
            "new_customers": new_customers,
            "active_customers": active_customers,
            "customer_lifetime_value": float(customer_ltv),
            "top_customers": [
                {
                    "id": customer_id,
                    "email": email,
                    "full_name": full_name,
                    "total_spent": float(total_spent),
                    "order_count": order_count,
                }
                for customer_id, email, full_name, total_spent, order_count in top_customers
            ],
            "customer_acquisition_trend": [
                {"year": int(year), "month": int(month), "new_customers": new_customers}
                for year, month, new_customers in customer_acquisition
            ],
        }

    async def get_product_analytics(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        tenant_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Get product performance analytics"""

        if not end_date:
            end_date = date.today()
        if not start_date:
            start_date = end_date - timedelta(days=30)

        # Base filters for orders
        order_filters = [
            Order.created_at >= start_date,
            Order.created_at <= end_date + timedelta(days=1),
            Order.status == OrderStatus.COMPLETED,
        ]

        if tenant_id:
            order_filters.append(Order.tenant_id == tenant_id)

        # Top selling products
        top_products = (
            self.db.query(
                Product.id,
                Product.name,
                Product.price,
                func.sum(OrderItem.quantity).label("total_sold"),
                func.sum(OrderItem.subtotal).label("total_revenue"),
                func.count(func.distinct(Order.id)).label("order_count"),
            )
            .join(OrderItem)
            .join(Order)
            .filter(*order_filters)
            .group_by(Product.id, Product.name, Product.price)
            .order_by(desc("total_revenue"))
            .limit(10)
            .all()
        )

        # Product categories performance
        category_performance = (
            self.db.query(
                Product.category,
                func.sum(OrderItem.quantity).label("total_sold"),
                func.sum(OrderItem.subtotal).label("total_revenue"),
                func.count(func.distinct(Product.id)).label("product_count"),
            )
            .join(OrderItem)
            .join(Order)
            .filter(*order_filters)
            .group_by(Product.category)
            .order_by(desc("total_revenue"))
            .all()
        )

        # Product conversion rates (views to purchases)
        # This would require tracking product views - placeholder for now
        conversion_data = []

        return {
            "top_selling_products": [
                {
                    "id": product_id,
                    "name": name,
                    "price": float(price),
                    "total_sold": total_sold,
                    "total_revenue": float(total_revenue),
                    "order_count": order_count,
                }
                for product_id, name, price, total_sold, total_revenue, order_count in top_products
            ],
            "category_performance": [
                {
                    "category": category,
                    "total_sold": total_sold,
                    "total_revenue": float(total_revenue),
                    "product_count": product_count,
                }
                for category, total_sold, total_revenue, product_count in category_performance
            ],
            "conversion_rates": conversion_data,
        }

    async def get_payment_analytics(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        tenant_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Get payment method and transaction analytics"""

        if not end_date:
            end_date = date.today()
        if not start_date:
            start_date = end_date - timedelta(days=30)

        filters = [
            Order.created_at >= start_date,
            Order.created_at <= end_date + timedelta(days=1),
        ]

        if tenant_id:
            filters.append(Order.tenant_id == tenant_id)

        # Payment method distribution
        payment_distribution = (
            self.db.query(
                Order.payment_method,
                func.count(Order.id).label("transaction_count"),
                func.sum(Order.total_amount).label("total_amount"),
                func.avg(Order.total_amount).label("avg_amount"),
            )
            .filter(*filters)
            .group_by(Order.payment_method)
            .all()
        )

        # Payment success rates
        success_rates = (
            self.db.query(
                Order.payment_method,
                func.count(case([(Order.status == OrderStatus.COMPLETED, 1)])).label(
                    "successful"
                ),
                func.count(case([(Order.status == OrderStatus.FAILED, 1)])).label(
                    "failed"
                ),
                func.count(Order.id).label("total"),
            )
            .filter(*filters)
            .group_by(Order.payment_method)
            .all()
        )

        # Daily transaction volume
        daily_transactions = (
            self.db.query(
                func.date(Order.created_at).label("date"),
                func.count(Order.id).label("transaction_count"),
                func.sum(Order.total_amount).label("total_amount"),
            )
            .filter(*filters)
            .group_by(func.date(Order.created_at))
            .order_by(func.date(Order.created_at))
            .all()
        )

        return {
            "payment_method_distribution": [
                {
                    "payment_method": method.value if method else "unknown",
                    "transaction_count": count,
                    "total_amount": float(total),
                    "average_amount": float(avg),
                }
                for method, count, total, avg in payment_distribution
            ],
            "success_rates": [
                {
                    "payment_method": method.value if method else "unknown",
                    "successful_transactions": successful,
                    "failed_transactions": failed,
                    "total_transactions": total,
                    "success_rate": (successful / total * 100) if total > 0 else 0,
                }
                for method, successful, failed, total in success_rates
            ],
            "daily_transaction_volume": [
                {
                    "date": date_val.isoformat(),
                    "transaction_count": count,
                    "total_amount": float(amount),
                }
                for date_val, count, amount in daily_transactions
            ],
        }

    async def get_geographic_analytics(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        tenant_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Get geographic distribution analytics"""

        if not end_date:
            end_date = date.today()
        if not start_date:
            start_date = end_date - timedelta(days=30)

        filters = [
            Order.created_at >= start_date,
            Order.created_at <= end_date + timedelta(days=1),
            Order.status == OrderStatus.COMPLETED,
        ]

        if tenant_id:
            filters.append(Order.tenant_id == tenant_id)

        # Revenue by country (assuming billing_country field exists)
        # This is a placeholder - would need to add geography fields to Order model
        country_revenue = []

        # City distribution
        city_revenue = []

        return {
            "revenue_by_country": country_revenue,
            "revenue_by_city": city_revenue,
            "geographic_insights": {"top_countries": [], "emerging_markets": []},
        }

    async def generate_executive_summary(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        tenant_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Generate executive summary dashboard"""

        # Get all analytics
        revenue_data = await self.get_revenue_analytics(start_date, end_date, tenant_id)
        customer_data = await self.get_customer_analytics(
            start_date, end_date, tenant_id
        )
        product_data = await self.get_product_analytics(start_date, end_date, tenant_id)
        payment_data = await self.get_payment_analytics(start_date, end_date, tenant_id)

        # Calculate key metrics
        total_orders = len(revenue_data.get("daily_revenue_trend", []))

        return {
            "summary": {
                "total_revenue": revenue_data["total_revenue"],
                "total_customers": customer_data["total_customers"],
                "new_customers": customer_data["new_customers"],
                "average_order_value": revenue_data["average_order_value"],
                "revenue_growth": revenue_data["revenue_growth_percentage"],
            },
            "top_metrics": {
                "best_selling_product": (
                    product_data["top_selling_products"][0]
                    if product_data["top_selling_products"]
                    else None
                ),
                "preferred_payment_method": (
                    payment_data["payment_method_distribution"][0]
                    if payment_data["payment_method_distribution"]
                    else None
                ),
                "top_customer": (
                    customer_data["top_customers"][0]
                    if customer_data["top_customers"]
                    else None
                ),
            },
            "trends": {
                "revenue_trend": revenue_data["daily_revenue_trend"][
                    -7:
                ],  # Last 7 days
                "customer_acquisition": customer_data["customer_acquisition_trend"][
                    -3:
                ],  # Last 3 months
            },
            "period": revenue_data["period"],
        }
